device/cuda/src/details/sparse_ccl.cu

```diff
@@ -0,0 +1,343 @@
+#include "edm/cell.hpp"
+#include "edm/measurement.hpp"
+
+#include "vecmem/memory/cuda/managed_memory_resource.hpp"
+#include "vecmem/memory/binary_page_memory_resource.hpp"
+#include "vecmem/memory/memory_resource.hpp"
+#include "vecmem/containers/vector.hpp"
+
+#include "sparse_ccl.cuh"
+
+#include <iostream>
+
+namespace traccc { namespace cuda { namespace details {
+    __device__
+    bool is_adjacent(const cell & a, const cell & b) {
+        return (
+            (a.channel0 - b.channel0)*(a.channel0 - b.channel0) <= 1 &&
+            (a.channel1 - b.channel1)*(a.channel1 - b.channel1) <= 1
+        );
+    }
+
+    __global__
+    void
+    sparse_ccl_kernel(
+        const cell * _cells,
+        const unsigned int * blocks,
+        float * _out
+    ) {
+        /*
+         * Select the relevant output region for this block. Since we are
+         * writing 4 floats per block (module) and we have a maximum of
+         * MAX_CLUSTERS_PER_MODULE output clusters per block, we need to seek
+         * four times that length into the array for each block.
+         */
+        float * out = &_out[4 * MAX_CLUSTERS_PER_MODULE * blockIdx.x];
+
+        /*
+         * Seek the correct cell region in the input data. Again, this is all a
+         * contiguous block of memory for now, and we use the blocks array to
+         * define the different ranges per block/module. At the end of this we
+         * have the starting address of the block of cells dedicated to this
+         * module, and we have its size.
+         */
+        const cell * cells = &_cells[blocks[blockIdx.x]];
+        const unsigned int size = blocks[blockIdx.x + 1] - blocks[blockIdx.x];
```

Comment:
Does `blocks[blockIdx.x + 1]` not go out of bounds for the last block?

Reply:
The blocks vector is defined as:

```cpp
vecmem::vector<unsigned int> blocks(data.headers.size() + 1, &mem);
```

So it's one larger than the number of blocks. :slightly_smiling_face: